xrtl/ui/control.h

```diff
@@ -0,0 +1,282 @@
+  // Opaque platform handle.
+  using PlatformHandle = uintptr_t;
+
+  // State of the control.
+  enum class State {
+    // Control is currently being created and will soon transition to the
+    // kCreated state. It does not yet have a usable platform handle.
+    kCreating,
+    // Control is created and has a platform handle available for use.
+    kCreated,
+
+    // Control is currently being destroyed and will soon transition to
+    // the kDestroyed state.
+    kDestroying,
+    // Control is destroyed.
+    // It cannot be used as there is no corresponding platform handle.
+    kDestroyed,
+  };
+
+  // Control event listener interface.
+  // These methods can be overridden to handle control-specific events.
+  //
+  // All listener callbacks occur on the message loop associated with the
+  // control.
+  struct Listener {
+    // Handles errors in the window system.
+    // If this is received the control is likely broken and must be destroyed.
+    virtual void OnError(ref_ptr<Control> target) {}
+
+    // Handles the control beginning to be created.
+    // It does not yet have a platform handle that is usable.
+    virtual void OnCreating(ref_ptr<Control> target) {}
+    // Handles the control ending creation.
+    // It now has a usable platform handle.
+    virtual void OnCreated(ref_ptr<Control> target) {}
+
+    // Handles the control beginning to be destroyed.
+    virtual void OnDestroying(ref_ptr<Control> target) {}
+    // Handles the control ending destruction.
+    // It no longer has a valid platform handle.
+    virtual void OnDestroyed(ref_ptr<Control> target) {}
+
+    // Handles the user changing the system theme (metrics/colors/etc).
+    virtual void OnSystemThemeChanged(ref_ptr<Control> target) {}
+
+    // Handles the control suspend state changing to the given new value.
+    virtual void OnSuspendChanged(ref_ptr<Control> target, bool is_suspended) {}
+    // Handles the control focus changing to the given new value.
+    virtual void OnFocusChanged(ref_ptr<Control> target, bool is_focused) {}
+
+    // Handles the control resizing to the given new bounds.
+    virtual void OnResized(ref_ptr<Control> target, Rect2D bounds) {}
+  };
+  using ListenerPtr = std::unique_ptr<Listener, void (*)(Listener*)>;
+
+  // Creates a new control using the given message loop for event dispatch.
+  static ref_ptr<Control> Create(ref_ptr<MessageLoop> message_loop);
+
+  // TODO(benvanik): API for wrapping existing controls by platform handle.
+
+  virtual ~Control() = default;
+
+  // Message loop the control is using for event dispatch.
+  ref_ptr<MessageLoop> message_loop() const { return message_loop_; }
+
+  // Platform handle of the control (such as an HWND) if the control is created.
+  // Implementation:
+  //  Android: ANativeWindow
+  //  iOS: UIView
+  //  MacOS: ?
+  //  Windows: HWND
+  //  X11: ::Window
+  virtual PlatformHandle platform_handle() = 0;
+
+  // Sets a listener that will receive event notifications.
```

Comment:
Yep - keeps things simple and usage contained. Thread-safe multi-listener event dispatch pretty much required copies of the listener list on each dispatch and I wanted to avoid that unless we really needed it. Looking at our current uses in the old code I don't think we do.